src/CrateTokenV1.sol

```diff
@@ -54,100 +44,123 @@ contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard {
+                uint256 crowdfundPrice = getBuyPrice(_amount - excess);
+                tokensInCurve -= (_amount - excess);
+                uint256 bondingCurvePrice = getBuyPrice(excess);
+                tokensInCurve -= excess;
+                crateFee = (crowdfundPrice / 10) + (bondingCurvePrice * CRATE_FEE_PERCENT) / 1 ether;
+                artistFee = ((crowdfundPrice * 9) / 10) + (bondingCurvePrice * ARTIST_FEE_PERCENT) / 1 ether;
+                totalPayment = crateFee + artistFee + bondingCurvePrice;
+            } else {
+                uint256 price = getBuyPrice(_amount);
+                tokensInCurve -= _amount;
+                crateFee = (price / 10);
+                artistFee = ((price * 9) / 10);
+                totalPayment = crateFee + artistFee;
+            }
+        } else if (phase == Phase.BONDING_CURVE) {
+            uint256 price = getBuyPrice(_amount);
+            tokensInCurve -= _amount;
+            crateFee = (price * CRATE_FEE_PERCENT) / 1 ether;
+            artistFee = (price * ARTIST_FEE_PERCENT) / 1 ether;
+            totalPayment = price + crateFee + artistFee;
+        } else revert WrongPhase();
 
-        tokensInCurve -= _amount;
-        _transfer(address(this), sender, _amount);
+        artistFees += artistFee;
+        if (msg.value < totalPayment) revert InsufficientPayment();
+        
+        _transfer(address(this), msg.sender, _amount);
+        emit TokenTrade(msg.sender, _amount, true, totalPayment);
 
-        if (tokensInCurve == 0) {
-            bondingCurveActive = false;
-            emit BondingCurveEnded();
-        }
-        emit TokenTrade(sender, _amount, true, totalPayment);
+        (bool refundSuccess,) = (msg.sender).call{value: msg.value - totalPayment}("");
+        if (!refundSuccess) revert TransferFailed();
 
         (bool crateFeePaid,) = protocolFeeDestination.call{value: crateFee}("");
-        require(crateFeePaid, "Failed to pay crate fee");
+        if (!crateFeePaid) revert TransferFailed();
 
-        // Refund the remaining Ether to the buyer
-        (bool refundSuccess,) = sender.call{value: msg.value - totalPayment}("");
-        require(refundSuccess, "Refund failed");
-
-        if (!bondingCurveActive) {
+        if (tokensInCurve == 0) {
+            phase = Phase.MARKET;
             _addLiquidity();
+            emit BondingCurveEnded();
         }
     }
 
     function sell(uint256 _amount, uint256 minEtherReceivable) external nonReentrant {
-        address sender = LibMulticaller.sender();
+        if (phase != Phase.BONDING_CURVE) revert WrongPhase();
+        if (balanceOf(msg.sender) < _amount) revert InsufficientTokens();
+        if (_amount < 10 ** decimals()) revert MustSellAtLeastOneToken();
+
+        /// @dev cache to use below for transfer
+        uint256 amount = _amount;
+        if (tokensInCurve + _amount >= CROWDFUND_THRESHOLD) {
+            phase = Phase.CROWDFUND;
```

Comment:
We should never be able to go back to the Crowdfund phase

Reply:
How should sales in the bonding curve phase be handled if they'd push the tokens held by the contract back into the crowd fund phase?